horovod/mxnet/mpi_ops.cc

```diff
@@ -0,0 +1,411 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <chrono>
+#include <memory>
+#include <thread>
+
+#include "../common/operations.h"
+#include "adapter.h"
+#include "cuda_util.h"
+#include "handle_manager.h"
+#include "mpi_ops.h"
+#include "ready_event.h"
+#include "tensor_util.h"
+
+namespace horovod {
+namespace mxnet {
+
+static HandleManager handle_manager;
+
+namespace {
+
+std::string GetOpName(std::string prefix, char* name) {
+  if (name != nullptr) {
+    return prefix + "." + std::string(name);
+  }
+  return std::string();
+}
+
+std::string GetOpNameHandle(std::string prefix, const std::string& name, int handle) {
+  if (name.length() != 0) {
+    return name;
+  }
+  return prefix + ".noname." + std::to_string(handle);
+}
+
+} // namespace
+
+int DoAllreduce(NDArray* tensor, NDArray* output, int average, const std::string& name, Callback cb) {
+  ThrowIfError(common::CheckInitialized());
+ 
+  auto handle = handle_manager.AllocateHandle(cb);
+  auto device = TensorUtil::GetDevice(tensor);
+  auto hvd_tensor = std::make_shared<MXTensor<NDArray>>(tensor);
+  auto hvd_context = std::make_shared<MXOpContext<NDArray>>(device, output);
+  auto hvd_output = std::make_shared<MXTensor<NDArray>>(output);
+
+  auto enqueue_result = EnqueueTensorAllreduce(
+      hvd_context, hvd_tensor, hvd_output, nullptr,
+      GetOpNameHandle("allreduce", name, handle), device,
+      [handle, average, output](const Status& status) {
+        handle_manager.MarkDone(handle, status);
+        handle_manager.ExecuteCallback(handle);
```

Comment:
Why don't just do `cb()` here?  Do they need to be synchronized to each other as you do in `ExecuteCallback`?

Reply:
Yeah, we were seeing segfaults when they were not synchronized with each other. @apeforest and @yuxihu tracked the segfaults to calling `cb()` here.

The specific commit: https://github.com/uber/horovod/commit/b2ef69eefbffb93bf61562bccf85af2164d56f22#diff-141f1184be39395e8f8708ff51a6fba7

Reply:
Can we just introduce a helper function inside this file that does locking and runs callback, and use it directly w/o need for handle manager?  Would we miss any functionality this way?